### PR TITLE
Bug fix: add error observation to history

### PR DESCRIPTION
**Pull request by li-boxuan**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

#1776 introduced a regression:

```python
    async def add_error_to_history(self, message: str):
        await self.add_history(NullAction(), ErrorObservation(message))
```
was replaced with

```python
    async def report_error(self, message: str):
        await self.event_stream.add_event(ErrorObservation(message), EventSource.AGENT)
```

and thus error observations are no longer added to history.

The negative implication is LLM might take more turns to fix the error, or even never able to fix the error. One example can be found [here](https://github.com/OpenDevin/OpenDevin/pull/2609#discussion_r1650352446). In short, LLM responds in wrong format multiple times but never gets any feedback.

**Give a brief summary of what the PR does, explaining any non-trivial design decisions**

One-liner fix: add error observation to history, so that the LLM can fix the error.

From the updated test artifacts, we can see now the LLM is able to fix the error in one fewer turn. This might be coincidence, but you can at least see from `tests/integration/mock/ManagerAgent/test_simple_task_rejection/prompt_004.log` that the prompt now contains the latest error observation. Prior to this PR, `prompt_003.log`, `prompt_004.log`, `prompt_005.log`, `prompt_006.log` are exactly the same.

**Other references**

### CODE REVIEW

File: opendevin/controller/agent_controller.py [hunk 1]
```diff
@@ -185,6 +185,9 @@ async def on_event(self, event: Event):
             elif isinstance(event, AgentDelegateObservation):
                 await self.add_history(NullAction(), event)
                 logger.info(event, extra={'msg_type': 'OBSERVATION'})
+            elif isinstance(event, ErrorObservation):
```
Comment (SmartManoj) [inline opendevin/controller/agent_controller.py:188]: How about adding the `else` part too and filtering unwanted events?

File: opendevin/controller/agent_controller.py [hunk 2]
```diff
@@ -185,6 +185,9 @@ async def on_event(self, event: Event):
...
+                await self.add_history(NullAction(), event)
+                logger.info(event, extra={'msg_type': 'OBSERVATION'})
```
Comment (SmartManoj) [inline opendevin/controller/agent_controller.py:191]: How about adding the else part too and filtering unwanted events?

**Comment by enyst** [comment]
Interesting find. You found it with delegation, but as far as I can tell, nothing about it is delegate-specific.

**Review by enyst — APPROVED** [review]

File: opendevin/controller/agent_controller.py (hunk 2 above)
Comment (li-boxuan) [reply opendevin/controller/agent_controller.py:191]: Please correct me if I misunderstood, but I think what you are suggesting here is, doing 

```python
await self.add_history(NullAction(), event)
logger.info(event, extra={'msg_type': 'OBSERVATION'})
```

for every type of observation by default, and use a block-list to skip events we don't want to handle like this.

If that's what you mean, my answer is, it makes sense, but it's safer to do it this way (allow-list), unless we know most observations should be added to history.

**Pull request merged**